### PR TITLE
OPHJOD-812: Add yksilön suosikki API

### PR DESCRIPTION
**Pull request by markus-boman-elenium**
## Description

Implementation of API to handle favorites of the yksilö.

image

## Related JIRA ticket

https://jira.eduuni.fi/browse/OPHJOD-812

### CODE REVIEW

File: src/main/java/fi/okm/jod/yksilo/dto/profiili/SuosikkiDto.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+public record SuosikkiDto(
+    @Schema(
+            name = "id",
+            requiredMode = Schema.RequiredMode.NOT_REQUIRED,
+            accessMode = Schema.AccessMode.READ_ONLY)
+        Optional<UUID> id,
+    @Schema(name = "suosionKohdeId", requiredMode = Schema.RequiredMode.REQUIRED) @NotNull
+        UUID suosionKohdeId,
+    @Schema(
```
Comment (jhyoty) [inline src/main/java/fi/okm/jod/yksilo/dto/profiili/SuosikkiDto.java:27]: Schema is redundant (notnull already makes the property required and the name (and example) is inferred from the field name and type).

File: src/main/java/fi/okm/jod/yksilo/dto/profiili/SuosikkiDto.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+            name = "tyyppi",
+            example = "TYOMAHDOLLISUUS",
+            requiredMode = Schema.RequiredMode.REQUIRED)
+        @NotNull
+        SuosikkiTyyppi tyyppi,
+    @Schema(
+            name = "luotu",
+            requiredMode = Schema.RequiredMode.NOT_REQUIRED,
+            accessMode = Schema.AccessMode.READ_ONLY)
+        Optional<LocalDateTime> luotu) {}
```
Comment (jhyoty) [inline src/main/java/fi/okm/jod/yksilo/dto/profiili/SuosikkiDto.java:37]: This use of optional is unnecessary and perhaps even misleading. When adding a suosikki, "luotu" can not be present. And when fetching a suosikki, luotu is always present. The type should be "Instant" and not localdatetime.

File: src/main/java/fi/okm/jod/yksilo/dto/profiili/SuosikkiDto.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import java.time.LocalDateTime;
+import java.util.Optional;
+import java.util.UUID;
+
+public record SuosikkiDto(
+    @Schema(
+            name = "id",
+            requiredMode = Schema.RequiredMode.NOT_REQUIRED,
+            accessMode = Schema.AccessMode.READ_ONLY)
+        Optional<UUID> id,
```
Comment (jhyoty) [inline src/main/java/fi/okm/jod/yksilo/dto/profiili/SuosikkiDto.java:24]: Unnecessary use of optional (a nullable field works here equally well and avoids extra ceremony (Optional.ofNullable(..)) on usage sites.

+1 for using Schema.accessMode, we should apply it consistently to other readonly properties (ids) as well.

File: src/main/java/fi/okm/jod/yksilo/entity/YksilonSuosikki.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.UUID;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import org.hibernate.annotations.Check;
+
+@Entity
+@Getter
+@Builder
```
Comment (jhyoty) [inline src/main/java/fi/okm/jod/yksilo/entity/YksilonSuosikki.java:35]: While builder is handy, I think it introduces a bit too much magic (code generation) and we do not use it elsewhere. In addition, this class has only handful of mandatory fields so writing a constructor is simple (and I'm not sure if you can express invariant like tyomahdollisuus xor koulutusmahdollisuus using a builder.

File: src/test/resources/data/create/yksilon-suosikki-test-data.sql [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+Työpaikat sijaitsevat muoviteollisuuden yrityksissä eri puolilla Suomea. Esimerkkejä työnantajista ovat Coreplast, Orthex, MSK Plast, Prevex, Kaivotuote ja Talokaivo.
+
+Työ edellyttää usein valmiutta vuorotyöhön, kuten kolmivuorotyöhön. Työ voi olla fyysisesti raskasta ja edellyttää hyvää peruskuntoa. Tärkeiksi taidoiksi mainitaan näppäryys, tarkkuus, huolellisuus, oma-aloitteisuus ja oppimiskyky. Aikaisempi työkokemus muoviteollisuudesta tai vastaavista tehtävistä katsotaan eduksi, mutta ei ole välttämätöntä, sillä työhön perehdytetään.
+
+Työnantajat tarjoavat vakituisia ja määräaikaisia työsuhteita, joissa noudatetaan muoviteollisuuden työehtosopimusta. Palkkaus määräytyy työkokemuksen ja osaamisen mukaan, mutta on yleensä työehtosopimuksen mukainen tai hieman korkeampi.
+
+Muovituotantotyöntekijä, muovihitsaaja, kokoonpanija, operaattori, asentaja, maalaamotyöntekijä, raaka-ainevarastotyöntekijä, tuotantotyöntekijä.');
```
Comment (jhyoty) [inline src/test/resources/data/create/yksilon-suosikki-test-data.sql:25]: Perhaps the test data could be simpler (no need to have the description)

File: src/main/java/fi/okm/jod/yksilo/entity/YksilonSuosikki.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+@Table(indexes = {@Index(columnList = "yksilo_id")})
+// TODO: move constraint to migration tool and add koulutusmahdollisuus type when available
+@Check(constraints = "tyomahdollisuus_id IS NOT NULL AND tyyppi = 'TYOMAHDOLLISUUS'")
+@AllArgsConstructor
+@NoArgsConstructor
+public class YksilonSuosikki {
+  @GeneratedValue @Id private UUID id;
+
+  @Column(name = "luotu", nullable = false, updatable = false)
+  private LocalDateTime createdAt;
```
Comment (jhyoty) [inline src/main/java/fi/okm/jod/yksilo/entity/YksilonSuosikki.java:45]: Should be "Instant", localdatetime is "timestamp without time zone"

**Review by jhyoty — CHANGES_REQUESTED** [review]

File: src/main/java/fi/okm/jod/yksilo/dto/profiili/SuosikkiDto.java (hunk 1 above)
Comment (markus-boman-elenium) [reply src/main/java/fi/okm/jod/yksilo/dto/profiili/SuosikkiDto.java]: Changed.

File: src/main/java/fi/okm/jod/yksilo/dto/profiili/SuosikkiDto.java (hunk 2 above)
Comment (markus-boman-elenium) [reply src/main/java/fi/okm/jod/yksilo/dto/profiili/SuosikkiDto.java]: Use of optional removed and replaced with nullable.

File: src/main/java/fi/okm/jod/yksilo/entity/YksilonSuosikki.java (hunk 4 above)
Comment (markus-boman-elenium) [reply src/main/java/fi/okm/jod/yksilo/entity/YksilonSuosikki.java]: Builder removed and replaced with tyomahdollisuus specific constructor.

File: src/test/resources/data/create/yksilon-suosikki-test-data.sql (hunk 5 above)
Comment (markus-boman-elenium) [reply src/test/resources/data/create/yksilon-suosikki-test-data.sql]: Data simplified.

File: src/main/java/fi/okm/jod/yksilo/entity/YksilonSuosikki.java (hunk 6 above)
Comment (markus-boman-elenium) [reply src/main/java/fi/okm/jod/yksilo/entity/YksilonSuosikki.java]: Changed to Instant.

**Review by jhyoty — APPROVED** [review]

**Pull request merged**